torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -399,121 +401,114 @@ def _step_microbatches(
         """
         arg_mbs, kwarg_mbs = self._check_inputs(arg_mbs, kwarg_mbs, target_mbs, losses)
 
-        # Example, 4 GPUs, 8 microbatches
-        # Stage 0: 6 warmup, 2 1f1b, 6 cooldown
-        # Stage 1: 4 warmup, 4 1f1b, 4 cooldown
-        # Stage 2: 2 warmup, 6 1f1b, 2 cooldown
-        # Stage 3: 0 warmup, 8 1f1b, 0 cooldown
-        # fwd only
-        warmup_steps = min(
+        # Last stage has 1 warmup, second-to-last 2 warmups, ...
+        # first stage `num_stages` warmups
+        warmup_chunks = min(
             self._n_microbatches,
-            2 * (self._num_stages - self._stage.stage_index - 1),
-        )
-        # fwd + bwd
-        main_1f1b_steps = self._n_microbatches - warmup_steps
-        # bwd only
-        cooldown_steps = (2 * self._n_microbatches) - (
-            warmup_steps + (2 * main_1f1b_steps)
-        )
-        total_steps = warmup_steps + main_1f1b_steps + cooldown_steps
-        logger.debug(
-            f"Stage {self._stage.stage_index}: "  # noqa: G004
-            f"Warmup steps: {warmup_steps}, "
-            f"Main 1F1B steps: {main_1f1b_steps}, "
-            f"Cooldown steps: {cooldown_steps}, "
-            f"Total steps: {total_steps}"
+            self._num_stages - self._stage.stage_index,
         )
 
-        # Delay send waits
-        fwd_sends_to_wait: List[dist.Work] = []
-        bwd_sends_to_wait: List[dist.Work] = []
-
-        def step_has_forward(i):
-            assert i >= 0, i
-            return i < self._n_microbatches
-
-        def step_has_backward(i):
-            assert i < total_steps, i
-            return i >= warmup_steps and self._has_backward
-
-        def is_1f1b_step(i):
-            return step_has_forward(i) and step_has_backward(i)
-
-        def is_warmup_step(i):
-            return step_has_forward(i) and not step_has_backward(i)
-
-        def is_cooldown_step(i):
-            return not step_has_forward(i) and step_has_backward(i)
-
-        def should_coalesce_fwd_send_bwd_recv(step):
-            return (
-                is_1f1b_step(step)
-                or (is_warmup_step(step) and is_cooldown_step(step + 1))
-                or (step >= 1 and is_warmup_step(step - 1) and is_cooldown_step(step))
-            )
-
-        def should_coalesce_bwd_send_fwd_recv(bwd_send_step):
-            # The backward send to prev stage should be coalesced with the fwd recv from the previous stage
-            return bwd_send_step >= warmup_steps and is_1f1b_step(bwd_send_step + 1)
-
-        # bwd chunk counter
+        # Chunk counters
+        fwd_mb_index = 0
         bwd_mb_index = 0
-        self._stage._configure_data_parallel_mode(last_backward=False)
```

Comment:
we need to add back the _configure_data_parallel_mode() call

Reply:
hm, isnt it better to add it back here still so its not broken in the mean time? 

Also does the PipelineStage api have enough info to know whether its the last backward?